File: questPresentation/src/main/java/ru/nekit/android/qls/utils/Declension.java
package ru.nekit.android.qls.utils;

import android.support.annotation.NonNull;

public class Declension {

    public static String declineAdjectiveByNoun(@NonNull String adjectiveBase,
                                                @NonNull String nounBase,
                                                @NonNull String format,
                                                @NonNull Gender gender,
                                                boolean isPlural) {
        return String.format(format, declineAdjective(adjectiveBase, gender, isPlural),
                declineNoun(nounBase, gender, isPlural));
    }

    public static String declineNoun(@NonNull String base, @NonNull Gender gender, boolean isPlural) {
        return base + (gender == Gender.NEUTER ? "о" : (isPlural ? "и" : "у"));
    }

    private static String declineAdjective(@NonNull String base, @NonNull Gender gender, boolean isPlural) {
        String ending = "";
        if (isPlural) {

            switch (gender) {

                case MALE:

                    ending = "ые";

                    break;

                case FEMALE:

                    ending = "ые";

                    break;

                case NEUTER:

                    ending = "ые";

                    break;

            }

        } else {

            switch (gender) {

                case MALE:

                    ending = "ой";

                    break;

                case FEMALE:

                    ending = "ую";

                    break;

                case NEUTER:

                    ending = "ое";

                    break;

            }
        }
        return base + ending;
    }

    public enum Gender {
        NEUTER,
        MALE,
        FEMALE
    }

}

File: questDomain/src/main/java/ru/nekit/android/qls/domain/utils/Declension.kt
package ru.nekit.android.qls.domain.utils

import ru.nekit.android.qls.domain.utils.Declension.Gender.*

object Declension {

    fun declineAdjectiveByNoun(adjectiveBase: String,
                               nounBase: String,
                               format: String,
                               gender: Gender,
                               isPlural: Boolean): String {
        return String.format(format, declineAdjective(adjectiveBase, gender, isPlural),
                declineNoun(nounBase, gender, isPlural))
    }

    fun declineNoun(base: String, gender: Gender, isPlural: Boolean): String {
        return base + if (gender == NEUTER) "о" else if (isPlural) "и" else "у"
    }

    private fun declineAdjective(base: String, gender: Gender, isPlural: Boolean): String = base + if (isPlural) {
        when (gender) {
            MALE -> "ые"
            FEMALE -> "ые"
            NEUTER -> "ые"
        }
    } else {
        when (gender) {
            MALE -> "ой"
            FEMALE -> "ую"
            NEUTER -> "ое"
        }
    }

    enum class Gender {
        NEUTER,
        MALE,
        FEMALE
    }

}
